### ISSUE TITLE
spring.cloud.config.server.bootstrap bug

### ISSUE THREAD

**Issue by xyq000** [spring-cloud/spring-cloud-config#897]
I wrote a spring cloud config server, whose `bootstrap.properties` was as follows:
```
spring.cloud.config.server.bootstrap=true
spring.application.name=yp-center
spring.profiles.active=native
```
However, enabling `spring.cloud.config.server.bootstrap` would incur a bug that the config server would use the configurations in its `application.properties` as the common external configs of all the client. This can be observed by visiting certain url, such as http://config-server-host:port/config-client/default

**Comment by ryanjbaxter** [comment]
So without `spring.cloud.config.server.bootstrap=true` this bug does not appear?

**Comment by spencergibb** [comment]
It might have to do with native

**Comment by ryanjbaxter** [comment]
Right but from what the documentation says it seems like we would expect all properties from `application.properties` to be removed.

**Comment by xyq000** [comment]
@ryanjbaxter yes, nothing goes wrong without `spring.cloud.config.server.bootstrap=true`. I have not tested that whether this bug appears when using non-native repository.

**Comment by ryanjbaxter** [comment]
This was a tough one to figure out the root cause.  It is a little complicated so here we go.

It all has to do with when `NativeRepositoryConfiguration` is created.  When `spring.cloud.config.server.bootstrap=true` `NativeRepositoryConfiguration` is created during the bootstrap phase (very early on when the application is started up) and before this line of code is run
https://github.com/spring-cloud/spring-cloud-commons/blob/v1.3.2.RELEASE/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapApplicationListener.java#L192
In this case `NativeRepositoryConfiguration` is created here https://github.com/spring-cloud/spring-cloud-commons/blob/v1.3.2.RELEASE/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapApplicationListener.java#L187
The reason it is created here is because `ConfigServerBootstrapConfiguration` imports `EnvironmentRepositoryConfiguration` which imports `NativeRepositoryConfiguration` and `ConfigServerBootstrapConfiguration` is only run when `spring.cloud.config.server.bootstrap=true` as you can see here https://github.com/spring-cloud/spring-cloud-config/blob/v1.4.2.RELEASE/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/bootstrap/ConfigServerBootstrapConfiguration.java#L44.  When `NativeEnvironmentRepository` is created this early on the `application.yml` has not yet been added as a property source to the environment.  So the `Environment` variable in `NativeEnvironmentRepository` does not contain that property source.  This means when the code in `NativeEnvironmentRepository.clean` runs it won't fine `application.yml` in the Environment and will include it in the response to the client.
https://github.com/spring-cloud/spring-cloud-config/blob/v1.4.2.RELEASE/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/NativeEnvironmentRepository.java#L214

When `NativeEnvironmentRepository` is not instantiated during the bootstrap phase, ie when `spring.cloud.config.server.bootstrap` is not set to `true`, then this line of code has a chance to run https://github.com/spring-cloud/spring-cloud-commons/blob/v1.3.2.RELEASE/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapApplicationListener.java#L192
What this will do is remove `spring.config.name=bootstrap` from the environment.  Then when this line of code runs in `ConfigFileApplicationListener`
https://github.com/spring-projects/spring-boot/blob/v1.5.10.RELEASE/spring-boot/src/main/java/org/springframework/boot/context/config/ConfigFileApplicationListener.java#L644
`application` will be returned and it will pickup `application.yml` and add it to the `PropertySources` in the `Environment`.  Now when `NativeEnvironmentRepository` is instantiated the `Environment` variable will contain the `application.yml` and will not include it when `clean` is run.

After thinking about this for a while I think this all makes sense given that when `spring.cloud.config.server.bootstrap=true` this will mean that the config server is looking to configure itself from the configured environment repository.  In order to do that it would need `application.yml` to be returned because it is a client of itself.  But because you are using a native repository and have not set any search locations its going to include what is on the classpath.  If you set a search location that did not include the classpath it would not include the `application.yml`.

I hope that all makes sense :)
